File: src/apple/Packages/MapleUI/Sources/MapleUI/OrganismsB/MuiBatchRenameModal.swift
// MuiBatchRenameModal.swift — Maple UI Organisms · Modals (unified-
// component-catalog.md §4.4). Template-driven rename with a live preview
// list, built on Overlay Shell from Form Field (template text), Chip Row
// (insertable tokens), Preview List, and Progress. The rename mapping is a
// pure function of `items`, `template`, and `startNumber`, recomputed on
// every keystroke — ported verbatim from the web reference's
// `applyTemplate`.

import SwiftUI

public struct MuiBatchRenameSourceItem: Identifiable, Sendable {
    public let id: String
    public let filename: String
    /// `YYYY-MM-DD` (or any ISO date string — only the leading 10
    /// characters are used) the `{date}` token formats from.
    public let date: String
    public let camera: String?

    public init(id: String, filename: String, date: String, camera: String? = nil) {
        self.id = id
        self.filename = filename
        self.date = date
        self.camera = camera
    }
}

public struct MuiBatchRenameResult: Sendable {
    public let template: String
    public let mapping: [MuiPreviewItem]
}

public struct MuiBatchRenameModal: View {
    private static let seqPadWidth = 3
    public static let tokens: [MuiChip] = [
        MuiChip(id: "date", label: "{date}"),
        MuiChip(id: "seq", label: "{seq}"),
        MuiChip(id: "camera", label: "{camera}"),
    ]

    public let isPresented: Bool
    public let contained: Bool
    public let items: [MuiBatchRenameSourceItem]
    @Binding public var template: String
    public let startNumber: Int
    public let renaming: Bool
    public let progress: Double
    public let renameConfirmed: ((MuiBatchRenameResult) -> Void)?
    public let dismissed: (() -> Void)?

    public init(
        isPresented: Bool,
        contained: Bool = false,
        items: [MuiBatchRenameSourceItem],
        template: Binding<String>,
        startNumber: Int = 1,
        renaming: Bool = false,
        progress: Double = 0,
        renameConfirmed: ((MuiBatchRenameResult) -> Void)? = nil,
        dismissed: (() -> Void)? = nil
    ) {
        self.isPresented = isPresented
        self.contained = contained
        self.items = items
        self._template = template
        self.startNumber = startNumber
        self.renaming = renaming
        self.progress = progress
        self.renameConfirmed = renameConfirmed
        self.dismissed = dismissed
    }

    private var previewItems: [MuiPreviewItem] {
        Self.previewItems(items: items, template: template, startNumber: startNumber)
    }

    public var body: some View {
        MuiOverlayShell(isPresented: isPresented, accessibilityLabel: "Batch Rename", contained: contained) {
            MuiText("Batch Rename", variant: .sheetTitle)
        } content: {
            VStack(alignment: .leading, spacing: MuiTokens.spacingMd) {
                MuiFormField(label: "Template", value: $template)
                MuiChipRow(chips: Self.tokens, mode: .select, selectedId: Binding(get: { nil }, set: { insertToken($0) }))
                MuiPreviewList(items: previewItems)
                if renaming {
                    MuiProgress(shape: .bar, value: progress, label: "\(Int(progress))%")
                }
            }
        } footer: {
            HStack {
                Spacer()
                MuiButton(label: "Cancel", variant: .ghost) { dismissed?() }
                MuiButton(label: "Rename", variant: .primary, isLoading: renaming, disabled: renaming) { confirmRename() }
            }
        } dismissed: {
            dismissed?()
        }
    }

    private func insertToken(_ tokenId: String?) {
        guard let tokenId, let token = Self.tokens.first(where: { $0.id == tokenId }) else { return }
        template += token.label
    }

    private func confirmRename() {
        renameConfirmed?(MuiBatchRenameResult(template: template, mapping: previewItems))
    }

    // MARK: - Pure logic (unit-testable without a live view)

    public static func previewItems(items: [MuiBatchRenameSourceItem], template: String, startNumber: Int) -> [MuiPreviewItem] {
        items.enumerated().map { index, item in
            MuiPreviewItem(id: item.id, before: item.filename, after: applyTemplate(template, item: item, seq: startNumber + index))
        }
    }

    static func applyTemplate(_ template: String, item: MuiBatchRenameSourceItem, seq: Int) -> String {
        let dateText = formattedDate(item.date)
        let seqText = String(format: "%0\(seqPadWidth)d", seq)
        return template
            .replacingOccurrences(of: "{date}", with: dateText)
            .replacingOccurrences(of: "{seq}", with: seqText)
            .replacingOccurrences(of: "{camera}", with: item.camera ?? "")
    }

    private static func formattedDate(_ iso: String) -> String {
        String(iso.prefix(10))
    }
}

#Preview("MuiBatchRenameModal") {
    struct Demo: View {
        @State private var open = false
        @State private var template = "{date}_{seq}"
        var body: some View {
            ZStack {
                MuiTokens.bg
                MuiButton(label: "Open Batch Rename", variant: .primary) { open = true }
                MuiBatchRenameModal(
                    isPresented: open,
                    items: [
                        MuiBatchRenameSourceItem(id: "1", filename: "IMG_0042.dng", date: "2026-08-01", camera: "SonyA7IV"),
                        MuiBatchRenameSourceItem(id: "2", filename: "IMG_0043.dng", date: "2026-08-01", camera: "SonyA7IV"),
                    ],
                    template: $template,
                    dismissed: { open = false }
                )
            }
            .frame(width: 420, height: 340)
        }
    }
    return Demo()
}

File: src/windows/Maple.WinUI/MapleUI/Organisms/MuiBatchRenameModal.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Maple.UI.Atoms;

namespace Maple.UI
{
    /// <summary>
    /// Maple.UI Batch Rename modal organism (unified-component-catalog.md
    /// §4.4, "Batch Rename" row: "Template with live preview", built from
    /// Form Field, Chip Row, Preview List, Progress) — a template
    /// <see cref="MuiInput"/> (Form Field), a Chip Row of insertable
    /// tokens, and a live <see cref="MuiPreviewList"/> of
    /// before/after names recomputed on every keystroke via
    /// <see cref="MuiBatchRenameLogic"/>.
    /// </summary>
    public sealed class MuiBatchRenameModal : ContentControl
    {
        private static readonly IReadOnlyList<MuiChip> TokenChips = new[]
        {
            new MuiChip("{name}", "{name}"), new MuiChip("{ext}", "{ext}"),
            new MuiChip("{date}", "{date}"), new MuiChip("{seq}", "{seq}"),
        };

        public static readonly DependencyProperty IsOpenProperty =
            DependencyProperty.Register(nameof(IsOpen), typeof(bool), typeof(MuiBatchRenameModal),
                new PropertyMetadata(false, (d, e) => ((MuiBatchRenameModal)d)._shell.IsOpen = (bool)e.NewValue));

        public static readonly DependencyProperty ContainedProperty =
            DependencyProperty.Register(nameof(Contained), typeof(bool), typeof(MuiBatchRenameModal),
                new PropertyMetadata(false, (d, e) => ((MuiBatchRenameModal)d)._shell.Contained = (bool)e.NewValue));

        public static readonly DependencyProperty TemplateProperty =
            DependencyProperty.Register(nameof(Template), typeof(string), typeof(MuiBatchRenameModal),
                new PropertyMetadata("{date}_{seq}", (d, _) => ((MuiBatchRenameModal)d).Rebuild()));

        public static readonly DependencyProperty OriginalsProperty =
            DependencyProperty.Register(nameof(Originals), typeof(IReadOnlyList<string>), typeof(MuiBatchRenameModal),
                new PropertyMetadata(null, (d, _) => ((MuiBatchRenameModal)d).Rebuild()));

        public static readonly DependencyProperty IsRenamingProperty =
            DependencyProperty.Register(nameof(IsRenaming), typeof(bool), typeof(MuiBatchRenameModal),
                new PropertyMetadata(false, (d, _) => ((MuiBatchRenameModal)d).Rebuild()));

        public bool IsOpen { get => (bool)GetValue(IsOpenProperty); set => SetValue(IsOpenProperty, value); }
        public bool Contained { get => (bool)GetValue(ContainedProperty); set => SetValue(ContainedProperty, value); }
        public string Template { get => (string)GetValue(TemplateProperty); set => SetValue(TemplateProperty, value); }

        /// <summary>Original filenames (with extension) being renamed.</summary>
        public IReadOnlyList<string>? Originals
        {
            get => (IReadOnlyList<string>?)GetValue(OriginalsProperty);
            set => SetValue(OriginalsProperty, value);
        }

        public bool IsRenaming { get => (bool)GetValue(IsRenamingProperty); set => SetValue(IsRenamingProperty, value); }

        public event EventHandler? Dismissed;
        public event EventHandler<string>? RenameRequested;

        private readonly MuiOverlayShell _shell = new() { Size = MuiOverlayShellSize.Md, AriaLabel = "Batch Rename" };
        private readonly MuiInput _templateInput = new() { Placeholder = "{date}_{seq}" };
        private readonly MuiChipRow _tokens = new() { Mode = MuiChipRowMode.Select, Chips = TokenChips };
        private readonly MuiPreviewList _preview = new();
        private readonly MuiProgress _progress = new() { IsIndeterminate = true, Visibility = Visibility.Collapsed };
        private readonly MuiButton _cancel = new() { Variant = MuiButtonVariant.Ghost, Label = "Cancel" };
        private readonly MuiButton _rename = new() { Variant = MuiButtonVariant.Primary, Label = "Rename all" };

        public MuiBatchRenameModal()
        {
            var body = new StackPanel { Orientation = Orientation.Vertical, Spacing = 14 };
            body.Children.Add(new MuiFormField { Label = "Template", ControlContent = _templateInput, Help = "Insert a token, or type your own." });
            body.Children.Add(_tokens);
            body.Children.Add(_preview);

            var footer = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 8, HorizontalAlignment = HorizontalAlignment.Right };
            footer.Children.Add(_progress);
            footer.Children.Add(_cancel);
            footer.Children.Add(_rename);

            _shell.Header = new MuiText { Text = "Batch Rename", Variant = MuiTextVariant.SheetTitle };
            _shell.Body = body;
            _shell.Footer = footer;
            Content = _shell;
            HorizontalContentAlignment = HorizontalAlignment.Stretch;
            VerticalContentAlignment = VerticalAlignment.Stretch;

            _shell.Dismissed += (_, _) => { IsOpen = false; Dismissed?.Invoke(this, EventArgs.Empty); };
            _cancel.Click += (_, _) => { IsOpen = false; Dismissed?.Invoke(this, EventArgs.Empty); };
            _templateInput.TextChanged += (_, text) => Template = text;
            _tokens.SelectionChanged += (_, tokenId) => Template += tokenId;
            _rename.Click += (_, _) => RenameRequested?.Invoke(this, Template);

            Rebuild();
        }

        private static (string Name, string Ext) Split(string filename)
        {
            var dot = filename.LastIndexOf('.');
            return dot < 0 ? (filename, string.Empty) : (filename[..dot], filename[(dot + 1)..]);
        }

        private void Rebuild()
        {
            if (_templateInput.Text != Template) _templateInput.Text = Template;
            _progress.Visibility = IsRenaming ? Visibility.Visible : Visibility.Collapsed;
            _rename.IsEnabled = !IsRenaming;

            var originals = Originals ?? Array.Empty<string>();
            var pairs = originals.Select(Split).ToList();
            var renamed = MuiBatchRenameLogic.PreviewBatch(Template, pairs, DateOnly.FromDateTime(DateTime.Today), 1, 3);
            _preview.Items = originals.Select((original, i) => new MuiPreviewItem(original, original, renamed[i])).ToList();
        }
    }
}
